docs/core/testing/unit-testing-best-practices.md

```diff
@@ -0,0 +1,515 @@
+```
+
+In this case, we are checking a property on the Fake (asserting against it), so in the above code snippet, the `mockOrder` is a Mock.
+
+**It's important to get this terminology correct. If you call your stubs "mocks", other developers are going to make false assumptions about your intent.**
+
+The main thing to remember about mocks versus stubs is that mocks are just like stubs, but you assert against the mock object, whereas you do not assert against a stub.
+
+## Best Practices
+
+### Arranging Your Tests
+Arrange, Act, Assert is a common pattern when unit testing. As the name implies, it consists of three main actions:
+- *Arrange* your objects, creating and setting them up as necessary.
+- *Act* on an object.
+- *Assert* that something is as expected.
+
+#### Why?
+- Clearly separates what is being tested from the *arrange* and *assert* steps.
+- Less chance to intermix assertions with "Act" code.
+
+#### Bad:
+```csharp
+public void IsValidWord_InputIsNull_ReturnsFalse()
+{
+    // Arrange
+    var glossary = new Glossary();
+
+    // Assert
+    Assert.False(glossary.IsValidWord(null));
+}
+```
+
+#### Better:
+```csharp
+public void IsValidWord_InputIsNull_ReturnsFalse()
+{
+    var glossary = new Glossary();
+    
+    var result = glossary.IsValidWord(null);
+
+    Assert.False(result);
+}
+```
+
+### Naming Your Tests
+The name of your test should consist of three parts:
+- The name of the method being tested.
+- The scenario under which it's being tested.
+- The expected behavior when the scenario is invoked.
+
+#### Why?
+- Naming standards are important because they explicitly express intent of the test.
+
+#### Bad:
+```csharp
+public void Test_Invalid()
+{
+    var glossary = new Glossary();
+
+    var result = glossary.IsValidWord(null);
+
+    Assert.False(result);
+}
+```
+
+#### Better:
+```csharp
+public void IsValidWord_InputIsNull_ReturnsFalse()
+{
+    var glossary = new Glossary();
+
+    var result = glossary.IsValidWord(null);
+
+    Assert.False(result);
+}
+```
+
+### Avoid Magic Strings
+Naming variables in unit tests is as important, if not more important, than naming variables in production code. Unit tests should not contain magic strings.
+
+#### Why?
+- Prevents the need for the reader of the test to inspect the production code in order to figure out what makes the value special.
+- Explicitly shows what you're trying to *prove* rather than trying to *accomplish*.
+
+#### Bad:
+```csharp
+public void TryParseWord_InputIsNumber_ReturnsInvalidInputErrorCode()
+{
+    var glossary = new Glossary();
+
+    glossary.TryParseWord("1", out var result)
+
+    Assert.Equal(-1, result);
+}
+```
+
+#### Better:
+```csharp
+public void TryParseWord_InputIsNumber_ReturnsInvalidInputErrorCode()
+{
+    var glossary = new Glossary();
+    const int INVALID_INPUT = -1;
+
+    glossary.TryParseWord("1", out var result)
+
+    Assert.Equal(INVALID_INPUT, result);
+}
+```
+
+### Avoid Logic in Tests
```

Comment:
I like the  way you've modulated the force of the different recommendations with "Avoid" and "Prefer" in these sections.